src/piece_picker.cpp

```diff
@@ -1060,21 +1060,43 @@ namespace libtorrent {
 
 		piece_pos& p = m_piece_map[index];
 		int const prev_priority = p.priority(this);
-		erase_download_piece(i);
-		int const new_priority = p.priority(this);
+		if (!blocks.empty())
+		{
+			auto const binfo = mutable_blocks_for_piece(*i);
+			for (int const block : blocks)
+			{
+				block_info& info = binfo[block];
+				TORRENT_ASSERT(info.state >= block_info::state_writing);
+				if (info.state == block_info::state_writing)
+					--i->writing;
+				else if (info.state == block_info::state_finished)
+					--i->finished;
+				info.peer = nullptr;
+				info.state = block_info::state_none;
+			}
+			//p.download_state = piece_pos::piece_downloading;
+			update_piece_state(i);
+		}
+
+		if (blocks.empty() || i->requested + i->finished + i->writing == 0)
```

Comment:
if this block is not executed because one of these counters is non-zero, isn't there a risk that this piece will just get left open indefinitely?
It's not clear to me why it's safe to remove the piece under this condition for a v1 torrent, but not a v2 torrent. I would expect it to either be a problem today, or be fine to run this code unconditionally.

Reply:
Calling `erase_download_piece` will cause the entire piece to be re-downloaded, which is exactly what we are trying to avoid when `blocks` is not empty. Block hashes are never known for a v1 torrent so the piece will always be erased, whereas for v2 torrents where block hashes are known we try to only reset the states of the failing blocks and otherwise leave the downloading piece as-is.

The counters check is assuming that at least one block has had its state reset to `state_none` so the piece should always be able to make progress by re-downloading those blocks.

Reply:
right, of course. typically we don't want to run this for v2 torrents, since we just reset the bad blocks